lib/class-dsl.php

```diff
@@ -48,27 +158,71 @@ public static function all_terms( string $taxonomy, string $field, array $values
 	/**
 	 * Build an exists DSL fragment.
 	 *
-	 * @param  string $field WP field.
+	 * @param string $field Field to check for existence.
+	 *
 	 * @return array DSL fragment.
 	 */
-	public static function exists( string $field ): array {
+	public function exists( string $field ): array {
 		return [
 			'exists' => [
-				'field' => self::$controller->map_field( $field ),
+				'field' => $this->map_field( $field ),
 			],
 		];
 	}
 
+	/**
+	 * Maps a field key to the Elasticsearch mapped field path.
+	 *
+	 * @param string $field The field key to map.
+	 *
+	 * @return string The mapped field reference.
+	 */
+	public function map_field( string $field ): string {
+		return $this->field_map[ $field ] ?? $field;
+	}
+
+	/**
+	 * Map a meta field. This will swap in the data type.
+	 *
+	 * @param  string $meta_key Meta key to map.
+	 * @param  string $type Data type to map.
+	 * @return string The mapped field.
+	 */
+	public function map_meta_field( string $meta_key, string $type = '' ): string {
+		if ( ! empty( $type ) ) {
+			return sprintf( $this->map_field( 'post_meta.' . $type ), $meta_key );
+		} else {
+			return sprintf( $this->map_field( 'post_meta' ), $meta_key );
+		}
+	}
+
+	/**
+	 * Map a taxonomy field. This will swap in the taxonomy name.
+	 *
+	 * @param  string $taxonomy Taxonomy to map.
+	 * @param  string $field Field to map.
+	 * @return string The mapped field.
+	 */
+	public function map_tax_field( string $taxonomy, string $field ): string {
+		if ( 'post_tag' === $taxonomy ) {
+			$field = str_replace( 'term_', 'tag_', $field );
+		} elseif ( 'category' === $taxonomy ) {
+			$field = str_replace( 'term_', 'category_', $field );
+		}
```

Comment:
Actually I see the filter elasticsearch_extensions_aggregation_taxonomy_field in class-taxonomy.php

Reply:
For custom taxonomies, it's just the normal term field. In the field map, categories and tags are handled in a special way, which this code handles. For example, a category will have a field name of `category_slug` whereas a custom taxonomy will just be `term_slug`. So if we are dealing with categories here, it uses the more specific `category_slug` rather than the generic `term_slug`.